<a href="https://colab.research.google.com/github/LadonV/Operation-Research/blob/main/Lab_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Simulation Assignment



In module 1, I proposed the problem of having 10 cases of water with 10 bottles in each case and 10 oz in each bottle.  One case is known to be filled with one ounce less than all the other bottles.  A former student proposed a random selection method for finding the case with the bottles that have less water.  Essentially you'll pull one bottle and test it.  If it is 9 ounces, you have found the case that is short.  Let's explore this random test.

1. Write a random function call that returns 10 ounces 9 out of 10 times and 9 ounces one out of ten.

2. What is the expected weight of a random pull?

3. Test your expected weight by running a simulation drawing at least 100 bottles and taking the average of the wieghts.  Does it agree with your theoretical result?

4. We are particuarly interested in how many pulls would be required to find the case that was short.  If you test one at a time, how many pulls do you expect before you find the case that is short?

5. Create a function that creates a sequence of pulling bottles and stops when you find the bottle with 9 ounces.  Test this function a bunch of times.  Do you notice anything odd when comparing it to your previous result?

6. Propose a modification to the coding you did in the first step that would improve the results.  Hint:  Consider a different data structure that would be more applicable to the real world case.

In [1]:
import random

def get_ounces():
    if random.random() < 0.9:
     return 10
    else:
      return 9

print(f"{get_ounces()} ounces")


10 ounces


In [2]:
(10*0.9)+(9*0.1)

9.9

Expected weight of a random pull in ounces

In [3]:
def get_ounces():
  if random.random() < 0.9:
    return 10
  else:
    return 9

l=[]

for j in range (100):
  l.append(get_ounces())


sum(l)/100

9.91

This does agree with my theoretical result as it is very close to the result 9.9

#4. Since the odds are 1/10 to find a one that is short I would expect to find one every 10 pulls.

In [4]:

def get_ounces():
    if random.random() < 0.9:
        return 10
    else:
        return 9

def simulate_pulls_until_nine():
    pull_count = 0
    while True:
        pull_count += 1
        if get_ounces() == 9:
            return pull_count

print(f"{simulate_pulls_until_nine():.2f}")

10.00


The only thing odd is that it doesn't give around 9-10 very often, haas very extreme outcomes

In [5]:

def get_ounces():
    if random.random() < 0.9:
        return 10
    else:
        return 9

def simulate_pulls_until_nine():
    pull_count = 0
    while True:
        pull_count += 1
        if get_ounces() == 9:
            return pull_count

pull_counts = []

for j in range(1000):
    pull_counts.append(simulate_pulls_until_nine())

average_pulls = sum(pull_counts) / 1000

print(f"{average_pulls:.2f}")

9.96


It takes around 10 times on average to pull a case that is short

In [6]:
import random

ounces_bucket = [10]*9 + [9]

def get_ounces():
    return random.choice(ounces_bucket)

print(f"{get_ounces()} ounces")

10 ounces


## Simulating Plates

In my household, there is a stack of plates that are differing in colors; pink, blue, black and green.  Of the 24 plates, 6 are each color.  My daughter will not eat dinner if it is not served on a pink plate.  

1.  Compute the probability that you draw a pink plate.
2. Compute the probability that of the four plates drawn for family dinner, atleast one of them is pink.
3. Create a randomization and test the previous two questions answers through randomization.  Discuss the results.
4. Let's make this a bit more complicated and indicative of how we live.  The plates are in a stack where the last one in, is the first one out.  Create a data structure that can account for this.
5. Also, we normally wash the plates when about 12 (sometimes more, sometimes less) collect in the dishwasher.  At that point, they are shuffled before returning to the stack in the cabinet.  Add this to your randomization and compute the probability of pulling exactly one and at least one pink plate when setting the table for four.
6. Since my daughter will not eat unless on the pink plate, most of the pink plates have migrated to the top of the stack.  This happened through a rejection of pull process.  In this, if we did not have at least one pink, more plates were removed from the stack until a pink was found.  Those rejected plates were then returned to the stack.  Create a function for doing this.
7. We have lived in the house for 1000 days.  Run your experiment on your stack always appeasing my daughter with at least one pink plate.  How often do you have all pink plates in the stack?
8. Explore what your final stack looks like by creating a usage count for each of the plates.  Have all the pink plates made it to the top and are some not in regular use?  

In [7]:
import math

probability = math.comb(6, 1) / math.comb(24, 1)
print(probability)


0.25


25% chance to draw a pink plate

In [8]:
zero_pink = math.comb(18, 4) / math.comb(24, 4)

at_least_one = 1 - zero_pink

print(at_least_one)

0.7120271033314511


71.2% chace for at least one of the 4 picked plates to be pink

In [9]:

plates = ['pink']*6 + ['nonpink']*18

pink_count_single = 0
pink_count_four = 0

for j in range(1000):
    random.shuffle(plates)

    draw1 = random.sample(plates, 1)
    if 'pink' in draw1:
        pink_count_single += 1

    draw4 = random.sample(plates, 4)
    if 'pink' in draw4:
        pink_count_four += 1

p_single = pink_count_single / 1000
p_four = pink_count_four / 1000

print(f"Estimated probability of drawing 1 pink plate: {p_single:.4f}")
print(f"Estimated probability of at least 1 pink in 4 draws: {p_four:.4f}")


Estimated probability of drawing 1 pink plate: 0.2500
Estimated probability of at least 1 pink in 4 draws: 0.7210


The estimated probability for question 1 proves that there is around a 25% chance to draw a pink plate over the other colored plates. The estimated probability for question 2 showes that there is around a 72 persent to pick a pink plate when grabbing 4 random plates. These results match up with the computed probabilities that were previously calculated.

In [10]:

plates_stack = ['pink']*6 + ['blue']*6 + ['black']*6 + ['green']*6

random.shuffle(plates_stack)

top_plate = plates_stack.pop()
print(f"Drew a {top_plate} plate")

drawn_plates = [plates_stack.pop() for _ in range(4)]
print(f"Drew these 4 plates: {drawn_plates}")

pink_single_count = 0
pink_four_count = 0

for j in range(1000):
    plates_stack = ['pink']*6 + ['blue']*6 + ['black']*6 + ['green']*6
    random.shuffle(plates_stack)

    if plates_stack.pop() == 'pink':
        pink_single_count += 1

    plates_stack = ['pink']*6 + ['blue']*6 + ['black']*6 + ['green']*6
    random.shuffle(plates_stack)
    drawn = [plates_stack.pop() for _ in range(4)]
    if 'pink' in drawn:
        pink_four_count += 1

p_single = pink_single_count / 1000
p_four = pink_four_count / 1000

print(f"Estimated probability of drawing 1 pink plate: {p_single:.4f}")
print(f"Estimated probability of at least 1 pink in 4 draws: {p_four:.4f}")


Drew a pink plate
Drew these 4 plates: ['blue', 'blue', 'blue', 'black']
Estimated probability of drawing 1 pink plate: 0.2930
Estimated probability of at least 1 pink in 4 draws: 0.7280


In [11]:

exactly_one_pink_count = 0
at_least_one_pink_count = 0

for j in range(1000):
    plates_stack = ['pink']*6 + ['nonpink']*18
    random.shuffle(plates_stack)

    drawn = [plates_stack.pop() for _ in range(4)]

    pink_in_draw = drawn.count('pink')

    if pink_in_draw == 1:
        exactly_one_pink_count += 1
    if pink_in_draw >= 1:
        at_least_one_pink_count += 1

    dishwasher_load = drawn[:min(12, len(drawn))]
    random.shuffle(dishwasher_load)

    plates_stack = plates_stack + dishwasher_load

p_exactly_one = exactly_one_pink_count / 1000
p_at_least_one = at_least_one_pink_count / 1000

print(f"Estimated probability of exactly one pink in 4 draws: {p_exactly_one:.4f}")
print(f"Estimated probability of at least one pink in 4 draws: {p_at_least_one:.4f}")


Estimated probability of exactly one pink in 4 draws: 0.4470
Estimated probability of at least one pink in 4 draws: 0.7040


In [12]:

def pull_until_pink(stack):
    drawn = []
    rejected = []

    while True:
        if not stack:
            break

        plate = stack.pop()
        drawn.append(plate)

        if plate == 'pink':
            break
        else:
            rejected.append(plate)

    stack.extend(rejected[::-1])

    return drawn, stack

plates_stack = ['pink']*6 + ['blue']*6 + ['black']*6 + ['green']*6
random.shuffle(plates_stack)

drawn, updated_stack = pull_until_pink(plates_stack)

print("Plates drawn for dinner:", drawn)
print("Top 5 plates remaining in stack:", updated_stack[-5:])



Plates drawn for dinner: ['black', 'blue', 'pink']
Top 5 plates remaining in stack: ['green', 'green', 'blue', 'blue', 'black']


In [13]:
import random

def pull_until_pink(stack):
    drawn = []
    rejected = []

    while True:
        if not stack:
            break
        plate = stack.pop()
        drawn.append(plate)
        if plate == 'pink':
            break
        else:
            rejected.append(plate)

    stack.extend(rejected[::-1])
    return drawn, stack

plates_stack = ['pink']*6 + ['blue']*6 + ['black']*6 + ['green']*6
random.shuffle(plates_stack)

days = 1000
all_pink_count = 0

for day in range(days):
    drawn, plates_stack = pull_until_pink(plates_stack)

    top_six = plates_stack[-6:] if len(plates_stack) >= 6 else plates_stack
    if all(p == 'pink' for p in top_six):
        all_pink_count += 1

prob_all_pink_top = all_pink_count / days
print(f"Estimated fraction of days with all pink plates at top: {prob_all_pink_top:.4f}")


Estimated fraction of days with all pink plates at top: 0.0000
